deps/rabbit/src/rabbit_deprecated_features.erl

```diff
@@ -0,0 +1,480 @@
+%% `DeprecatedFeature' is a {@type deprecated_feature_modattr()}.
+%%
+%% == How to check that a deprecated feature is permitted ==
+%%
+%% To check in the code if a deprecated feature is permitted:
+%%
+%% ```
+%% case rabbit_deprecated_features:is_permitted(DeprecatedFeatureName) of
+%%     true ->
+%%         %% The deprecated feature is still permitted.
+%%         ok;
+%%     false ->
+%%         %% The deprecated feature is gone or should be considered
+%%         %% unavailable.
+%%         error
+%% end.
+%% '''
+%%
+%% == How to allow or deny a deprecated feature from the configuration ==
+%%
+%% The following configuration snippet permits one deprecated feature and
+%% denies another one:
+%%
+%% ```
+%% permit_deprecated_features.my_deprecated_feature_1 = true
+%% permit_deprecated_features.my_deprecated_feature_2 = false
+%% '''
+
+-module(rabbit_deprecated_features).
+
+-include_lib("kernel/include/logger.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+-include_lib("rabbit_common/include/logging.hrl").
+
+-include("src/rabbit_feature_flags.hrl").
+
+-export([is_permitted/1,
+         get_phase/1,
+         get_warning/1]).
+-export([extend_properties/2,
+         should_be_permitted/2,
+         enable_underlying_feature_flag_cb/1]).
+
+-type deprecated_feature_modattr() :: {rabbit_feature_flags:feature_name(),
+                                       feature_props()}.
+%% The value of a `-rabbitmq_deprecated_feature()' module attribute used to
+%% declare a deprecated feature.
+%%
+%% Example:
+%% ```
+%% -rabbit_deprecated_feature(
+%%    {my_deprecated_feature_1,
+%%     #{deprecation_phase => permitted_by_default,
+%%       warning => "Feature 1 will be removed from RabbitMQ X.0"
+%%      }}).
+%% '''
+
+-type deprecation_phase() :: permitted_by_default |
+                             denied_by_default |
+                             disconnected |
+                             removed.
+%% The deprecation phase of a feature.
+%%
+%% Deprecation phases are used in the following order:
+%% <ol>
+%% <li>`permitted_by_default': the feature is enabled by default and the user
+%% can use it like they did so far. They can turn it off in the configuration
+%% to experiment with the absence of the feature.</li>
+%% <li>`denied_by_default': the feature is disabled by default and the user
+%% must enable it in the configuration to be able to use it.</li>
+%% <li>`disconnected': the code of the feature is still there but it is
+%% disabled and can't be re-enabled from the configuration. The user has to
+%% recompile RabbitMQ to re-enable it.</li>
+%% <li>`removed': the code of the feature is no longer in the product. There is
+%% no way to re-enable it at this point. The deprecated feature must still be
+%% defined because, like required feature flags, its presence is important to
+%% determine if nodes are compatible and can be clustered together.</li>
+%% </ol>
+
+-type feature_props() :: #{deprecation_phase := deprecation_phase(),
```

Comment:
I don't object to adding that, but I fail to see the benefit. To me, this is serves as documentation. Package pre-install scripts won't use this information, I don't know for orchestration tools.

A pre-install script would have to look at the permitted and actually used deprecated features in the currently installed and running instance (regardless of the deprecation phase of the deprecated features) and compare that to the deprecation phase of the package being installed. And "installed" here doesn't mean "upgraded". It could be reinstalled or downgraded.

Feature flags systems (not only in RabbitMQ) exist precisely to not have to guess things from versions. I hope orchestration systems don't take decisions based on versions.

Reply:
I've got this idea from the discussion in slack about cluster operator and how such metadata possibly can be saved and later extracted from docker image. 

And saving to a file is just for consistency - why not have it here if we are already providing it in docker metadata. Maybe some feature-rich tooling can be built by package maintainers - but yes, I also don't have any usecase for such a file in mind. Only suggested because it costs almost nothing to produce it.

(And I know that e.g. golang SDK always provides `VERSION` file in the root folder, so you can understand what's there without running any commands - so at least at some places a similar pattern is applicable).